internal/unionstore/art/arena.go

```diff
@@ -0,0 +1,358 @@
+	if !addr.isNull() {
+		return addr, data
+	}
+
+	a.enlarge(size, a.blockSize<<1)
+	addr, data = a.allocInLastBlock(size, align)
+	if prevBlocks != len(a.blocks) {
+		a.onMemChange()
+	}
+	return addr, data
+}
+
+func (a *memArena) enlarge(allocSize, blockSize int) {
+	a.blockSize = blockSize
+	for a.blockSize <= allocSize {
+		a.blockSize <<= 1
+	}
+	// Size will never larger than maxBlockSize.
+	if a.blockSize > maxBlockSize {
+		a.blockSize = maxBlockSize
+	}
+	a.blocks = append(a.blocks, memArenaBlock{
+		buf: make([]byte, a.blockSize),
+	})
+	a.capacity += uint64(a.blockSize)
+	// We shall not call a.onMemChange() here, since it will make the latest block empty, which breaks a precondition
+	// for some operations (e.g. revertToCheckpoint)
+}
+
+func (a *memArena) allocInLastBlock(size int, align bool) (nodeAddr, []byte) {
+	idx := len(a.blocks) - 1
+	offset, data := a.blocks[idx].alloc(size, align)
+	if offset == nullBlockOffset {
+		return nullAddr, nil
+	}
+	return nodeAddr{uint32(idx), offset}, data
+}
+
+func (a *memArena) onMemChange() {
+	hook := a.memChangeHook.Load()
+	if hook != nil {
+		(*hook)()
+	}
+}
+
+func (a *memArena) reset() {
+	for i := range a.blocks {
+		a.blocks[i].reset()
+	}
+	a.blocks = a.blocks[:0]
+	a.blockSize = 0
+	a.capacity = 0
+	a.onMemChange()
+}
+
+func (a *memArenaBlock) alloc(size int, align bool) (uint32, []byte) {
+	offset := a.length
+	if align {
+		// We must align the allocated address for node
+		// to make runtime.checkptrAlignment happy.
+		offset = (a.length + 7) & alignMask
+	}
+	newLen := offset + size
+	if newLen > len(a.buf) {
+		return nullBlockOffset, nil
+	}
+	a.length = newLen
+	return uint32(offset), a.buf[offset:newLen]
+}
+
+func (a *memArenaBlock) reset() {
+	a.buf = nil
+	a.length = 0
+}
+
+const memdbVlogHdrSize = 8 + 8 + 4
+
+type vlogHdr struct {
+	nodeAddr nodeAddr
+	oldValue nodeAddr
+	valueLen uint32
+}
+
+func (hdr *vlogHdr) store(dst []byte) {
+	cursor := 0
+	endian.PutUint32(dst[cursor:], hdr.valueLen)
+	cursor += 4
+	hdr.oldValue.store(dst[cursor:])
+	cursor += 8
+	hdr.nodeAddr.store(dst[cursor:])
+}
+
+func (hdr *vlogHdr) load(src []byte) {
+	cursor := 0
+	hdr.valueLen = endian.Uint32(src[cursor:])
+	cursor += 4
+	hdr.oldValue.load(src[cursor:])
+	cursor += 8
+	hdr.nodeAddr.load(src[cursor:])
+}
+
+func (f *artAllocator) allocValue(leafAddr nodeAddr, oldAddr nodeAddr, value []byte) nodeAddr {
+	size := memdbVlogHdrSize + len(value)
+	addr, data := f.vlogAllocator.alloc(size, false)
+	copy(data, value)
+	hdr := vlogHdr{leafAddr, oldAddr, uint32(len(value))}
+	hdr.store(data[len(value):])
+	addr.off += uint32(size)
+	return addr
+}
+
+func (f *artAllocator) getValue(valAddr nodeAddr) []byte {
+	hdrOff := valAddr.off - memdbVlogHdrSize
+	block := f.vlogAllocator.blocks[valAddr.idx].buf
+	valLen := endian.Uint32(block[hdrOff:])
+	if valLen == 0 {
+		return tombstone
+	}
+	valOff := hdrOff - valLen
+	return block[valOff:hdrOff:hdrOff]
```

Comment:
I know this line is copied from memdbArena. I'm just wondering why its capacity is specified as `hdrOff`

Reply:
See [Full slice expressions](https://go.dev/ref/spec#Full_slice_expressions), the capacity is set to `max - low`(`hdrOff - valOff` in our code).